hpx/parallel/algorithms/is_heap.hpp

```diff
@@ -0,0 +1,348 @@
+    /// \tparam RandIter    The type of the source iterators used (deduced).
+    ///                     This iterator type must meet the requirements of a
+    ///                     random access iterator.
+    /// \tparam Comp        The type of the function/function object to use
+    ///                     (deduced).
+    ///
+    /// \param policy       The execution policy to use for the scheduling of
+    ///                     the iterations.
+    /// \param first        Refers to the beginning of the sequence of elements
+    ///                     the algorithm will be applied to.
+    /// \param last         Refers to the end of the sequence of elements the
+    ///                     algorithm will be applied to.
+    /// \param comp         comp is a callable object. The return value of the
+    ///                     INVOKE operation applied to an object of type Comp,
+    ///                     when contextually converted to bool, yields true if
+    ///                     the first argument of the call is less than the
+    ///                     second, and false otherwise. It is assumed that comp
+    ///                     will not apply any non-constant function through the
+    ///                     dereferenced iterator.
+    ///
+    /// \a comp has to induce a strict weak ordering on the values.
+    ///
+    /// The application of function objects in parallel algorithm
+    /// invoked with an execution policy object of type
+    /// \a sequenced_policy execute in sequential order in the
+    /// calling thread.
+    ///
+    /// The application of function objects in parallel algorithm
+    /// invoked with an execution policy object of type
+    /// \a parallel_policy or \a parallel_task_policy are
+    /// permitted to execute in an unordered fashion in unspecified
+    /// threads, and indeterminately sequenced within each thread.
+    ///
+    /// \returns  The \a is_heap algorithm returns a \a hpx::future<bool>
+    ///           if the execution policy is of type \a sequenced_task_policy or
+    ///           \a parallel_task_policy and returns \a bool otherwise.
+    ///           The \a is_heap algorithm returns whether the range is max heap.
+    ///           That is, true if the range is max heap, false otherwise.
+    ///
+    template <typename ExPolicy, typename RandIter, typename Comp = detail::less,
+    HPX_CONCEPT_REQUIRES_(
+        execution::is_execution_policy<ExPolicy>::value &&
+        hpx::traits::is_iterator<RandIter>::value)
+    >
+    typename util::detail::algorithm_result<ExPolicy, bool>::type
+    is_heap(ExPolicy && policy, RandIter first, RandIter last,
+        Comp && comp = Comp())
+    {
+        static_assert(
+            (hpx::traits::is_random_access_iterator<RandIter>::value),
+            "Requires a random access iterator.");
+
+        typedef execution::is_sequential_execution_policy<ExPolicy> is_seq;
+
+        return detail::is_heap<RandIter>().call(
```

Comment:
In case of is_heap_until, I can't break the partition checking lower index than an index of element which breaks the heap. I can only break the partition checking higher index than it.
But in case of is_heap, when in any partition I find an element which breaks the heap, I can break all partitions.